### DISCUSSION TITLE
implement async io

### DISCUSSION THREAD

**Discussion by maddanio** [vlang/v#11919] [Ideas]
when I first looked at v it seemed like it could be a go alternative for me in implementing an rtsp proxy for my company, but when looking at the networking lib it seems all networking is actually blocking. sure this can be masked by spawning lots of threads, but in my case with 100s of connections that will lead to 100s of threads which in my experience is a lot slower than a proper scheduler. on of go's main features (despite horrible resource handling semantics and the fact it runs garbage collected on a vm) for me is that it seems to have a very high performance scheduler for network io.

is async io support considered? or will all coroutines actually run on separate threads for the time being? or will coroiutines be able to yield, but there will be no support to yield during network io?

**Comment by mcastorina** [comment]
V does not have a runtime (like Go), so for the time being coroutines will run on separate OS threads. I believe a runtime / scheduler is planned, but I don't see it getting implemented any time soon.

**Comment by maddanio** [comment]
Well the primary thing needed for efficient networking would be reactor style async io. A provided scheduler and coroutines (which are integrated with a reactor in go's goroutines) would be nice but not required. It could also be done with a library though. Doing sync io with threads doesnt scale at all.

**Comment by JalonSolov** [comment]
Actually, Linux does a pretty good job of managing more threads than you have available cores to run them on, so async io works just fine with what V has now.  It may not scale as far as "green threads", but it does work.  Tests have been run that proved running 200 threads per core is only about 15% slower than running "only" 20 threads per core.  These were tests with V code.  See `vlib/sync/bench/channel_bench_v.v`

**Comment by maddanio** [comment]
Hmm, I remain sceptical. In my case I am usually running on about 100 sockets, doing simultanous reads and writes on all of them. In sync io world this would mean 200 threads running select calls on the sockets for synchronisation and lots of additional synchronisation overhead both in code (in async world I can make sure all callbacks related to a single socket are run on the same thread) and resulting in additional lock contention. Also sceptical because just about every modern language these days has async io and co-routines. well, maybe I will try writing my own benchmark actually running on 100 sockets or take a deeper look at yours. and I still think v as a modern language should have async io :).

**Reply by danieldaeschle** [reply]
There should be no problem with 100 threads.
